tests/unit/stream_functional_test.cxx

```diff
@@ -0,0 +1,573 @@
+    CHK_OK( s2.getTestSm()->isSame( *s1.getTestSm() ) );
+    CHK_OK( s3.getTestSm()->isSame( *s1.getTestSm() ) );
+
+    // Re-enable stream mode
+    update_stream_params(pkgs, 500);
+    CHK_Z( activate_stream(s1, s2_addr) );
+    CHK_Z( append_log_in_stream_without_delivery(s1, s2_addr, 10, 10) );
+    // 10 pending requests, 10 commit requests
+    CHK_Z( drain_pending_reqs_queue(pkgs, s1, s2_addr, 20) );
+    CHK_OK( s2.getTestSm()->isSame( *s1.getTestSm() ) );
+    CHK_OK( s3.getTestSm()->isSame( *s1.getTestSm() ) );
+
+    print_stats(pkgs);
+
+    s1.raftServer->shutdown();
+    s2.raftServer->shutdown();
+    s3.raftServer->shutdown();
+
+    f_base->destroy();
+    return 0;
+}
+
+int activate_and_deactivate_stream_mode_test() {
+    reset_log_files();
+    ptr<FakeNetworkBase> f_base = cs_new<FakeNetworkBase>();
+
+    std::string s1_addr = "S1";
+    std::string s2_addr = "S2";
+
+    RaftPkg s1(f_base, 1, s1_addr);
+    RaftPkg s2(f_base, 2, s2_addr);
+    std::vector<RaftPkg*> pkgs = {&s1, &s2};
+
+    CHK_Z( launch_servers( pkgs ) );
+    CHK_Z( make_group( pkgs ) );
+    for (auto& entry: pkgs) {
+        RaftPkg* pp = entry;
+        raft_params param = pp->raftServer->get_current_params();
+        param.return_method_ = raft_params::async_handler;
+        pp->raftServer->update_params(param);
+    }
+
+    // Set stream mode params
+    update_stream_params(pkgs, 500);
+
+    // Append 1 log to enable stream
+    CHK_Z( activate_stream(s1, s2_addr) );
+
+    // Append 10 logs in stream mode, pending reqs > 1
+    CHK_Z( append_log_in_stream_without_delivery(s1, s2_addr, 10, 10) );
+    // 10 pending requests, 10 commit requests
+    CHK_Z( drain_pending_reqs_queue(pkgs, s1, s2_addr, 20) );
+
+    // Remove s2 and then add back
+    s1.dbgLog(" --- remove S2---");
+    s1.raftServer->remove_srv( s2.getTestMgr()->get_srv_config()->get_id() );
+
+    // Leave req/resp.
+    s1.fNet->execReqResp();
+    // Leave done, notify to peers.
+    s1.fNet->execReqResp();
+    // Notify new commit.
+    s1.fNet->execReqResp();
+    // Wait for bg commit for configuration change.
+    CHK_Z( wait_for_sm_exec(pkgs, COMMIT_TIMEOUT_SEC) );
+
+    // add S2 back to leader.
+    s1.dbgLog(" --- add back S2---");
+    s1.raftServer->add_srv( *(s2.getTestMgr()->get_srv_config()) );
+    s1.fNet->execReqResp();
+    // Send the entire snapshot.
+    do {
+        s1.fNet->execReqResp();
+    } while (s2.raftServer->is_receiving_snapshot());
+    // Commit.
+    s1.fNet->execReqResp();
+    CHK_Z( wait_for_sm_exec(pkgs, COMMIT_TIMEOUT_SEC) );
+    CHK_Z( s1.fNet->getNumPendingReqs(s2_addr) );
+
+    // Append two logs, in non-stream mode, it only generate 1 reqs
+    CHK_Z( append_log_in_stream_without_delivery(s1, s2_addr, 1, 1) );
+    CHK_Z( append_log_in_stream_without_delivery(s1, s2_addr, 1, 1) );
+
+    // Activate stream mode
+    // It need to re-install the snapshot because of config change, 18 reqs
```

Comment:
After S2 re-enter into the cluster, it will trigger a duplicate snapshot install, is that expected?

Reply:
Because in S2, at first it received snapshot at `10`, and had log from `11` to `14`.
```
2024-09-18T17:07:47.820_650-07:00 [159298] [INFO] snapshot idx 10 term 1 is successfully applied, log start 11 last idx 14  [handle_snapshot_sync.cxx:593, handle_snapshot_sync_req()]
```
after that, log `15` was created for configuration in S1
```
2024-09-18T17:07:47.820_938-07:00 [159303] [INFO] new configuration: log idx 15, prev log idx 14peer 1, DC ID 1, S1, voting member, 50
peer 2, DC ID 1, S2, voting member, 50 
my id: 1, leader: 1, term: 1    [handle_commit.cxx:929, reconfigure()]
```
since snapshot distance is 5, and number of reserved log is 0, it immediately did log compaction, up to 15
```
2024-09-18T17:07:47.820_962-07:00 [159303] [INFO] snapshot idx 15 log_term 1 created, compact the log store if needed   [handle_commit.cxx:662, on_snapshot_completed()]
2024-09-18T17:07:47.820_964-07:00 [159303] [INFO] log_store_ compact upto 15    [handle_commit.cxx:673, on_snapshot_completed()]
```
which means, there is no way for S2 to catch-up, hence snapshot was sent again.
```
2024-09-18T17:07:47.821_256-07:00 [159298] [INFO] trying to sync snapshot with last index 15 to peer 2, its last log idx 14 [handle_snapshot_sync.cxx:144, create_sync_snapshot_req()]
```

So if you want to avoid sending snapshot twice, set `reserved_log_items_` greater than `0` at the beginning of the test. Maybe `5` is appropriate.